pkg/mutation/mutation_function.go

```diff
@@ -0,0 +1,140 @@
+package mutation
+
+import (
+	"fmt"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/path/parser"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func Mutate(mutator Mutator, obj *unstructured.Unstructured) error {
+	return mutate(mutator, obj.Object, nil, 0)
+}
+
+func mutate(m Mutator, current interface{}, previous interface{}, depth int) error {
+	if len(m.Path().Nodes)-1 == depth {
+		return addValue(m, current, previous, depth)
+	}
+	pathEntry := m.Path().Nodes[depth]
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+		next, ok := current.(map[string]interface{})[pathEntry.(*parser.Object).Reference]
+		if !ok {
+			next = createMissingElement(m, current, previous, depth)
+		}
+		if err := mutate(m, next, current, depth+1); err != nil {
+			return err
+		}
+		return nil
+	case parser.ListNode:
+		elementFound := false
+		listPathEntry := pathEntry.(*parser.List)
+		glob := listPathEntry.Glob
+		key := listPathEntry.KeyField
+		for _, listElement := range current.([]interface{}) {
+			if glob {
+				if err := mutate(m, listElement, current, depth+1); err != nil {
+					return err
+				}
+				elementFound = true
+			} else if elementValue, ok := listElement.(map[string]interface{})[key]; ok {
+				if *listPathEntry.KeyValue == elementValue {
+					if err := mutate(m, listElement, current, depth+1); err != nil {
+						return err
+					}
+					elementFound = true
+				}
+			}
+		}
+		// If no matching element in the array was found in non Globbed list, create a new element
+		if !listPathEntry.Glob && !elementFound {
+			next := createMissingElement(m, current, previous, depth)
+			if err := mutate(m, next, current, depth+1); err != nil {
+				return err
+			}
+		}
+	default:
+		return fmt.Errorf("Unrecognized type: %v", t)
+	}
+	return nil
+}
+
+func addValue(m Mutator, current interface{}, previous interface{}, depth int) error {
+	pathEntry := m.Path().Nodes[depth]
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+		key := pathEntry.(*parser.Object).Reference
+		switch m.(type) {
+		case *AssignMetadataMutator:
+
+			if elementValue, found, _ := unstructured.NestedString(current.(map[string]interface{}), key); found {
+				log.Info("Mutated value already present", "field", key, "value", elementValue)
+				return nil
+			}
+		}
+		value, err := m.Value()
+		if err != nil {
+			return err
+		}
+		if err = unstructured.SetNestedField(current.(map[string]interface{}), value, key); err != nil {
+			log.Error(err, "Failed to mutate object", "field", key, "value", value)
+		}
+
+	case parser.ListNode:
+		return setListElementToValue(m, current, previous, depth)
+	}
+	return nil
+}
+
+func setListElementToValue(m Mutator, current interface{}, previous interface{}, depth int) error {
+	listPathEntry := m.Path().Nodes[depth].(*parser.List)
```

Comment:
We should consider safe type casting here, as we have not checked the type of `Nodes[depth]` in this function